lib/util/api.ts

```diff
@@ -14,190 +14,280 @@
+let emitSourceMetadata: (
+	sourceMetadata: SourceMetadata | Record<string, never>,
+) => void | undefined; // Record<string, never> means an empty object
 
 /**
  * @summary Terminate the child process
  */
-async function terminate(exitCode: number) {
-	ipc.disconnect(IPC_SERVER_ID);
+async function terminate(exitCode?: number) {
 	await cleanup(Date.now());
 	process.nextTick(() => {
 		process.exit(exitCode || SUCCESS);
 	});
 }
 
-/**
- * @summary Handle errors
- */
-async function handleError(error: Error) {
-	emit('error', toJSON(error));
-	await delay(DISCONNECT_DELAY);
-	await terminate(GENERAL_ERROR);
+// kill the process if no initila connections or heartbeat for X sec (default 10)
+function setTerminateTimeout() {
+	if (ETCHER_TERMINATE_TIMEOUT > 0) {
+		return setTimeout(() => {
+			console.log(
+				`no connections or heartbeat for ${ETCHER_TERMINATE_TIMEOUT} ms, terminating`,
+			);
+			terminate();
+		}, ETCHER_TERMINATE_TIMEOUT);
+	} else {
+		return null;
+	}
 }
 
-/**
- * @summary Abort handler
- * @example
- */
-const onAbort = async (exitCode: number) => {
-	log('Abort');
-	emit('abort');
-	await delay(DISCONNECT_DELAY);
-	await terminate(exitCode);
-};
-
-const onSkip = async (exitCode: number) => {
-	log('Skip validation');
-	emit('skip');
-	await delay(DISCONNECT_DELAY);
-	await terminate(exitCode);
-};
-
-ipc.connectTo(IPC_SERVER_ID, () => {
-	// Gracefully exit on the following cases. If the parent
-	// process detects that child exit successfully but
-	// no flashing information is available, then it will
-	// assume that the child died halfway through.
+// terminate the process cleanly on SIGINT
+process.once('SIGINT', async () => {
+	await terminate(SUCCESS);
+});
 
-	process.once('uncaughtException', handleError);
+// terminate the process cleanly on SIGTERM
+process.once('SIGTERM', async () => {
+	await terminate(SUCCESS);
+});
 
-	process.once('SIGINT', async () => {
-		await terminate(SUCCESS);
-	});
+let TerminateInterval = setTerminateTimeout();
 
-	process.once('SIGTERM', async () => {
-		await terminate(SUCCESS);
-	});
+type EmitLog = {
+	emit: (channel: string, message: object | string) => void;
+	log: (message: string) => void;
+};
 
-	// The IPC server failed. Abort.
-	ipc.of[IPC_SERVER_ID].on('error', async () => {
-		await terminate(SUCCESS);
-	});
+function setup(): Promise<EmitLog> {
+	return new Promise((resolve, reject) => {
+		wss.on('connection', (ws) => {
+			console.log('connection established... setting up');
+
+			/**
+			 * @summary Send a message to the IPC server
+			 */
+			function emit(type: string, payload?: object | string) {
+				ws.send(JSON.stringify({ type, payload }));
+				// ipc.of[IPC_SERVER_ID].emit("message", { type, payload });
+			}
+
+			/**
+			 * @summary Print logs and send them back to client
+			 */
+			function log(message: string) {
+				console.log(message);
+				emit('log', message);
+			}
+
+			/**
+			 * @summary Handle `errors`
```

Comment:
All these doc strings don’t seem to add much besides noise. Delete them? Unless you feel like there’s subtleties in each function that you need to surface as a comment, in which case by all means do so.

Reply:
I have a tendency to over-comment. It's like using your turn signal in the middle of the night, all alone on the highway; it doesn't do anything, but it's a reflex, and I prefer to do it when it's not needed rather than risk not doing it when it would be necessary.

Reply:
But I'm happy to remove them if you feel strongly about it :)

Reply:
I’m not offended to see them, but if optimizing for readability is a worthy goal then comments that don’t offer valuable information have no place in the code base.